### PR TITLE
ITF: PortGuard

### PR DESCRIPTION
**Pull request by MBoldyrev**
### Description of the Change

Meet integration_framework::PortGuard - a class that keeps track of ports you use in ITF tests! Just add an instance in your class where you allocate ports, and ask it for some in a given range. When your class dies (sorry), it will in its final moments mark the ports your instance requested as free.

### Benefits

It should make port reusing easier.

### Possible Drawbacks 

It uses locks for synchronization.

### CODE REVIEW

File: test/framework/integration_framework/port_guard.cpp [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+                     "PortGuard chose a port outside boundaries!");
+    instance_used_ports_.set(tested_port);
+    all_used_ports_.set(tested_port);
+    return tested_port;
+  }
+
+  PortGuard::PortType PortGuard::getPort(const PortType &min_value,
+                                         const PortType &max_value) {
+    boost::optional<PortType> opt_port = tryGetPort(min_value, max_value);
+    BOOST_VERIFY_MSG(
```
Comment (l4l) [inline test/framework/integration_framework/port_guard.cpp]: Why *_VERIFY_* used here and *_ASSERT_* at the function above?

File: test/module/test/framework/integration_framework/port_guard_test.cpp [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+#include <boost/noncopyable.hpp>
+#include "framework/integration_framework/port_guard.hpp"
+
+using PortGuard = integration_framework::PortGuard;
+
+TEST(PortGuardTest, AllPortsGetUsedAndNoOverlaps) {
+  static constexpr size_t kNumClients = 10; // first-run clients
+  static constexpr size_t kNumClientsKilled = 3; // then these die
+  static constexpr size_t kNumClientsResurrected = 5; // then these are added
+  static constexpr std::pair<size_t, size_t> kRange = {123, 456};
```
Comment (l4l) [inline test/module/test/framework/integration_framework/port_guard_test.cpp]: Would be nice to have exact amount of available ports as the required number, since it's a better corner case

File: test/framework/integration_framework/port_guard.cpp [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "framework/integration_framework/port_guard.hpp"
+
+#include <boost/assert.hpp>
+
+namespace integration_framework {
+
+  constexpr PortGuard::PortType PortGuard::kMaxPort;
```
[the remark this answers is not in the document]
Comment (MBoldyrev) [reply test/framework/integration_framework/port_guard.cpp]: It needs to be somewhere. If you mean there is a more appropriate place, please suggest it.

File: test/framework/integration_framework/port_guard.cpp (hunk 1 above)
Comment (MBoldyrev) [reply test/framework/integration_framework/port_guard.cpp]: Because this error condition can happen in runtime and it is not mitigated in any way except that macro. And the assertions above are just for testing and clarification purposes and the conditions checked there cannot fail because they are mitigated by the code logic (until it is not tampered with, that is the purpose of asserts).

File: test/module/test/framework/integration_framework/port_guard_test.cpp (hunk 2 above)
Comment (MBoldyrev) [reply test/module/test/framework/integration_framework/port_guard_test.cpp]: Just tried with the whole port range - it was running for 1 min 16 sec. Too long, I think.

Comment (l4l) [reply test/module/test/framework/integration_framework/port_guard_test.cpp]: Maybe reduce the range then?

File: test/module/test/framework/integration_framework/port_guard_test.cpp [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+          << "Port " << port << " was not given out to any client!";
+    }
+  };
+
+  // first run
+  run_new_clients(kNumClients);
+  check();
+
+  // kill some clients...
+  clients.resize(kNumClients - kNumClientsKilled);
```
[the remark this answers is not in the document]
Comment (nickaleks) [reply test/module/test/framework/integration_framework/port_guard_test.cpp]: If it is to check that ports are freed, maybe write it as a separate test

File: test/module/test/framework/integration_framework/port_guard_test.cpp [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+      while (true) {
+        auto port = client.port_guard.tryGetPort(kRange.first, kRange.second);
+        if (!port) {
+          break;
+        }
+        client.used_ports.set(*port);
+      }
+    };
+
+    clients.reserve(clients.size() + num_clients);
```
Comment (nickaleks) [inline test/module/test/framework/integration_framework/port_guard_test.cpp]: since you have not put anything into clients, clients.size() will be 0

File: test/module/test/framework/integration_framework/port_guard_test.cpp [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+  // first run
+  run_new_clients(kNumClients);
+  check();
+
+  // kill some clients...
+  clients.resize(kNumClients - kNumClientsKilled);
+  // ... and run some new clients
+  run_new_clients(kNumClientsResurrected);
+  check();
```
Comment (nickaleks) [inline test/module/test/framework/integration_framework/port_guard_test.cpp]: Using check() twice will be confusing since it is unclear which one causes test failure

File: test/framework/integration_framework/integration_test_framework.hpp [hunk 7]
```diff
@@ -373,6 +374,7 @@ namespace integration_framework {
     std::map<std::string, tbb::concurrent_queue<TxResponseType>>
         responses_queues_;
 
+    std::unique_ptr<PortGuard> port_guard_;
```
Comment (nickaleks) [inline test/framework/integration_framework/integration_test_framework.hpp]: Why keep it in a pointer?

File: test/module/test/framework/integration_framework/port_guard_test.cpp [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+#include <gtest/gtest.h>
+#include <boost/thread/barrier.hpp>
+#include <boost/noncopyable.hpp>
+#include "framework/integration_framework/port_guard.hpp"
+
+using PortGuard = integration_framework::PortGuard;
+
+TEST(PortGuardTest, AllPortsGetUsedAndNoOverlaps) {
+  static constexpr size_t kNumClients = 10; // first-run clients
```
Comment (nickaleks) [inline test/module/test/framework/integration_framework/port_guard_test.cpp]: static constexpr out of class context does not make much sense.

File: test/framework/integration_framework/integration_test_framework.hpp (hunk 7 above)
Comment (MBoldyrev) [reply test/framework/integration_framework/integration_test_framework.hpp]: It is to hide the implementation.
